<a href="https://colab.research.google.com/github/RizkyaSalsabila/Assignment-Sem5_ML_Rizkya-Salsabila/blob/main/praktikum_13/TG13_2341720056_Rizkya_Salsabila.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# PRAKTIKUM 1 - JST sederhana (2 layer) dengan forward pass dan backpropagation manual

## Eksekusi Kode Praktikum 1

In [2]:
import numpy as np

# Dataset untuk masalah XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
# Label XOR (0 jika sama, 1 jika beda)
y = np.array([[0],[1],[1],[0]])

# Jumlah neuron input, hidden, dan output
input_size = 2
hidden_size = 2
output_size = 1
lr = 0.1  # Learning rate

# Inisialisasi bobot layer 1 dengan nilai random
W1 = np.random.randn(input_size, hidden_size)
# Inisialisasi bias layer 1 dengan nol
b1 = np.zeros((1, hidden_size))
# Inisialisasi bobot layer 2 dengan nilai random
W2 = np.random.randn(hidden_size, output_size)
# Inisialisasi bias layer 2 dengan nol
b2 = np.zeros((1, output_size))

# Fungsi aktivasi sigmoid untuk menormalkan output 0–1
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

# Turunan sigmoid untuk proses backpropagation
def sigmoid_derivative(x):
    return x * (1 - x)

# Loop training sebanyak 10.000 iterasi
for epoch in range(10000):

    # Menghitung input ke hidden layer (z1)
    z1 = np.dot(X, W1) + b1
    # Menghitung aktivasi hidden layer (a1)
    a1 = sigmoid(z1)

    # Menghitung input ke output layer (z2)
    z2 = np.dot(a1, W2) + b2
    # Menghitung output akhir jaringan (a2)
    a2 = sigmoid(z2)

    # Menghitung error antara label dan prediksi
    error = y - a2

    # Gradien output layer
    d_a2 = error * sigmoid_derivative(a2)
    # Gradien bobot W2
    d_W2 = np.dot(a1.T, d_a2)
    # Gradien bias b2
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    # Gradien hidden layer
    d_a1 = np.dot(d_a2, W2.T) * sigmoid_derivative(a1)
    # Gradien bobot W1
    d_W1 = np.dot(X.T, d_a1)
    # Gradien bias b1
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot W1, W2 dan bias b1, b2
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    # Cetak loss setiap 1000 epoch
    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss}")

# Cetak hasil prediksi akhir jaringan
print("Prediksi:")
print(a2)

Epoch 0, Loss: 0.25275039259698706
Epoch 1000, Loss: 0.24862955872670592
Epoch 2000, Loss: 0.2336537841397576
Epoch 3000, Loss: 0.08410007229585693
Epoch 4000, Loss: 0.02020295265005958
Epoch 5000, Loss: 0.009690546680615536
Epoch 6000, Loss: 0.006108439769457885
Epoch 7000, Loss: 0.004384885589915651
Epoch 8000, Loss: 0.0033907288622123033
Epoch 9000, Loss: 0.0027502892542091255
Prediksi:
[[0.04852543]
 [0.95435581]
 [0.94560466]
 [0.04276033]]


## Soal Praktikum 1

### SOAL 1 : Ubah jumlah neuron hidden layer menjadi 3

In [3]:
import numpy as np

# Dataset untuk masalah XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
# Label XOR (0 jika sama, 1 jika beda)
y = np.array([[0],[1],[1],[0]])

# Jumlah neuron input, hidden, dan output
input_size = 2
hidden_size = 3   # DIUBAH SESUAI SOAL = 3
output_size = 1
lr = 0.1  # Learning rate

# Inisialisasi bobot layer 1 dengan nilai random
W1 = np.random.randn(input_size, hidden_size)
# Inisialisasi bias layer 1 dengan nol
b1 = np.zeros((1, hidden_size))
# Inisialisasi bobot layer 2 dengan nilai random
W2 = np.random.randn(hidden_size, output_size)
# Inisialisasi bias layer 2 dengan nol
b2 = np.zeros((1, output_size))

# Fungsi aktivasi sigmoid untuk menormalkan output 0–1
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

# Turunan sigmoid untuk proses backpropagation
def sigmoid_derivative(x):
    return x * (1 - x)

# Loop training sebanyak 10.000 iterasi
for epoch in range(10000):

    # Menghitung input ke hidden layer (z1)
    z1 = np.dot(X, W1) + b1
    # Menghitung aktivasi hidden layer (a1)
    a1 = sigmoid(z1)

    # Menghitung input ke output layer (z2)
    z2 = np.dot(a1, W2) + b2
    # Menghitung output akhir jaringan (a2)
    a2 = sigmoid(z2)

    # Menghitung error antara label dan prediksi
    error = y - a2

    # Gradien output layer
    d_a2 = error * sigmoid_derivative(a2)
    # Gradien bobot W2
    d_W2 = np.dot(a1.T, d_a2)
    # Gradien bias b2
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    # Gradien hidden layer
    d_a1 = np.dot(d_a2, W2.T) * sigmoid_derivative(a1)
    # Gradien bobot W1
    d_W1 = np.dot(X.T, d_a1)
    # Gradien bias b1
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot W1, W2 dan bias b1, b2
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    # Cetak loss setiap 1000 epoch
    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss}")

# Cetak hasil prediksi akhir jaringan
print("Prediksi:")
print(a2)

Epoch 0, Loss: 0.27922107216847375
Epoch 1000, Loss: 0.2500818117191574
Epoch 2000, Loss: 0.2500108308056744
Epoch 3000, Loss: 0.2499707221860559
Epoch 4000, Loss: 0.249900957437653
Epoch 5000, Loss: 0.24971175465134227
Epoch 6000, Loss: 0.24898651619627613
Epoch 7000, Loss: 0.24388634255080008
Epoch 8000, Loss: 0.1906030395684695
Epoch 9000, Loss: 0.05267834772304581
Prediksi:
[[0.1028376 ]
 [0.85365826]
 [0.8654002 ]
 [0.15068381]]


### SOAL 2 : Bandingkan hasil loss dengan konfigurasi awal

Dari percobaan di atas, dapat dibandingkan hasilnya bahwa :
" Pada konfigurasi awal (2 neuron), loss turun dengan cepat dan mencapai nilai akhir 0.00275.
Pada konfigurasi hidden 3 neuron, loss turun lebih lambat dan hanya mencapai 0.0527. Artinya, konfigurasi awal lebih baik, karena jaringan lebih stabil dan lebih cepat menemukan pola XOR. "

### SOAL 3 : Tambahkan fungsi aktivasi ReLU dan bandingkan hasil.

In [5]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Arsitektur
input_size = 2
hidden_size = 3   # hidden neuron = 3 (sesuai soal)
output_size = 1
lr = 0.1

# Inisialisasi bobot & bias
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# === FUNGSI AKTIVASI ===
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

# --- ReLU untuk hidden layer ---
def relu(x):
    return np.maximum(0, x)

def relu_derivative(x):
    return (x > 0).astype(float)

# ============================
# TRAINING
# ============================
for epoch in range(10000):

    # Forward Pass
    z1 = np.dot(X, W1) + b1
    a1 = relu(z1)            # GANTI SIGMOID → RELU

    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)         # Output tetap sigmoid

    # Error
    error = y - a2

    # Backprop Output Layer
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    # Backprop Hidden Layer
    d_a1 = np.dot(d_a2, W2.T) * relu_derivative(a1)   # DERIVATIVE RELU
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update Bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    # Print monitoring loss
    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss}")

# Output akhir
print("Prediksi:")
print(a2)

Epoch 0, Loss: 0.25008150258581013
Epoch 1000, Loss: 0.1668396683343036
Epoch 2000, Loss: 0.16674472509081645
Epoch 3000, Loss: 0.16671724962084816
Epoch 4000, Loss: 0.1667019068793097
Epoch 5000, Loss: 0.1666945844244845
Epoch 6000, Loss: 0.16668899014776367
Epoch 7000, Loss: 0.16668832080092136
Epoch 8000, Loss: 0.16668409304457088
Epoch 9000, Loss: 0.16668384905971667
Prediksi:
[[0.33342049]
 [0.33342049]
 [0.99240146]
 [0.33342049]]


Setelah menambahkan fungsi aktivasi ReLU pada hidden layer, performa jaringan meningkat. Loss yang semula stagnan pada 0.25 (menggunakan sigmoid) berhasil turun hingga 0.166. Prediksi output juga menjadi lebih mendekati nilai target XOR dibanding konfigurasi awal. Hal ini menunjukkan bahwa ReLU lebih efektif dalam mempercepat konvergensi dan menghindari vanishing gradient pada jaringan neural sederhana ini.